File: packages/sdk/pact-ts/src/session/session.ts
import {
  deleteSession,
  postNotification,
  postRpc,
  terminalMessage,
  type JsonRpcMessage,
  type RpcExchange,
  type RpcMessageHandler,
  type SessionState,
} from "../transport/index.ts";
import { isJsonRpcFailure, type JsonRpcNotification, type JsonRpcRequest, type JsonRpcResponse } from "../types.ts";

export interface PactSessionHandshake {
  initializeResponse: RpcExchange;
  initializedResponse: RpcExchange;
}

export interface PactSessionOptions extends SessionState {
  authToken: string;
  baseUrl: string;
  fetchImpl?: typeof fetch;
  onMessage?: RpcMessageHandler;
  handshake?: PactSessionHandshake | null;
}

export class PactSession {
  readonly authToken: string;
  readonly baseUrl: string;
  handshake: PactSessionHandshake | null;
  readonly protocolVersion: string;
  readonly sessionId: string;

  #fetchImpl: typeof fetch;
  #nextRequestId: number;
  #onMessage: RpcMessageHandler;

  constructor(options: PactSessionOptions) {
    this.authToken = options.authToken;
    this.baseUrl = options.baseUrl;
    this.handshake = options.handshake ?? null;
    this.protocolVersion = options.protocolVersion;
    this.sessionId = options.sessionId;
    this.#fetchImpl = options.fetchImpl ?? fetch;
    this.#nextRequestId = 2;
    this.#onMessage = options.onMessage ?? (async () => {});
  }

  setMessageHandler(onMessage: RpcMessageHandler): void {
    this.#onMessage = onMessage;
  }

  async request<TParams = unknown>(
    method: string,
    params?: TParams,
    onMessage: RpcMessageHandler = this.#onMessage,
  ): Promise<RpcExchange> {
    const request: JsonRpcRequest<TParams> = {
      jsonrpc: "2.0",
      id: this.#nextId(),
      method,
    };
    if (params !== undefined) {
      request.params = params;
    }

    return postRpc(
      this.baseUrl,
      this.authToken,
      this.sessionId,
      this.protocolVersion,
      request,
      onMessage,
      this.#fetchImpl,
    );
  }

  async sendEnvelope(
    body: Record<string, unknown>,
    onMessage: RpcMessageHandler = this.#onMessage,
  ): Promise<RpcExchange> {
    return postRpc(
      this.baseUrl,
      this.authToken,
      this.sessionId,
      this.protocolVersion,
      body,
      onMessage,
      this.#fetchImpl,
    );
  }

  async requestResult<TResult = unknown, TParams = unknown>(
    method: string,
    params?: TParams,
    onMessage: RpcMessageHandler = this.#onMessage,
  ): Promise<JsonRpcResponse<TResult>> {
    const exchange = await this.request(method, params, onMessage);
    return terminalMessage(
      exchange.messages,
      exchange.request.id as string | number | null,
    ) as unknown as JsonRpcResponse<TResult>;
  }

  async notification<TParams = unknown>(
    method: string,
    params?: TParams,
    onMessage: RpcMessageHandler = this.#onMessage,
  ): Promise<RpcExchange> {
    const notification: JsonRpcNotification<TParams> = {
      jsonrpc: "2.0",
      method,
    };
    if (params !== undefined) {
      notification.params = params;
    }

    return postNotification(
      this.baseUrl,
      this.authToken,
      this.sessionId,
      this.protocolVersion,
      notification,
      onMessage,
      this.#fetchImpl,
    );
  }

  async listTools(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("tools/list", params);
  }

  async callTool(name: string, args: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("tools/call", {
      name,
      arguments: args,
    });
  }

  async listResources(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("resources/list", params);
  }

  async readResource(uri: string): Promise<unknown> {
    return this.#result("resources/read", { uri });
  }

  async subscribeResource(uri: string): Promise<unknown> {
    return this.#result("resources/subscribe", { uri });
  }

  async unsubscribeResource(uri: string): Promise<unknown> {
    return this.#result("resources/unsubscribe", { uri });
  }

  async listResourceTemplates(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("resources/templates/list", params);
  }

  async listPrompts(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("prompts/list", params);
  }

  async getPrompt(name: string, args?: Record<string, unknown>): Promise<unknown> {
    const params: Record<string, unknown> = { name };
    if (args) {
      params.arguments = args;
    }
    return this.#result("prompts/get", params);
  }

  async complete(params: Record<string, unknown>): Promise<unknown> {
    return this.#result("completion/complete", params);
  }

  async setLogLevel(level: string): Promise<RpcExchange> {
    return this.notification("logging/setLevel", { level });
  }

  async listTasks(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.#result("tasks/list", params);
  }

  async getTask(taskId: string): Promise<unknown> {
    return this.#result("tasks/get", { taskId });
  }

  async getTaskResult(taskId: string): Promise<unknown> {
    return this.#result("tasks/result", { taskId });
  }

  async cancelTask(taskId: string): Promise<unknown> {
    return this.#result("tasks/cancel", { taskId });
  }

  async close(): Promise<{ status: number; headers: Record<string, string> }> {
    return deleteSession(this.baseUrl, this.authToken, this.sessionId, this.#fetchImpl);
  }

  #nextId(): number {
    const id = this.#nextRequestId;
    this.#nextRequestId += 1;
    return id;
  }

  async #result<TResult = unknown>(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<TResult> {
    const response = await this.requestResult<TResult>(method, params);
    if (isJsonRpcFailure(response)) {
      throw new Error(response.error.message);
    }
    return response.result;
  }
}

File: packages/sdk/chio-go/session/session.go
package session

import (
	"context"
	"fmt"
	"net/http"
	"sync/atomic"

	"github.com/backbay/chio/packages/sdk/chio-go/transport"
)

type MessageHandler = transport.MessageHandler
type RPCExchange = transport.RPCExchange
type DeleteSessionResult = transport.DeleteSessionResult

type SessionHandshake struct {
	InitializeResponse  RPCExchange
	InitializedResponse RPCExchange
}

type Session struct {
	AuthToken       string
	BaseURL         string
	Handshake       *SessionHandshake
	HTTPClient      *http.Client
	ProtocolVersion string
	SessionID       string

	nextRequestID atomic.Int64
	onMessage     MessageHandler
}

type Options struct {
	AuthToken       string
	BaseURL         string
	Handshake       *SessionHandshake
	HTTPClient      *http.Client
	OnMessage       MessageHandler
	ProtocolVersion string
	SessionID       string
}

func New(options Options) *Session {
	instance := &Session{
		AuthToken:       options.AuthToken,
		BaseURL:         options.BaseURL,
		Handshake:       options.Handshake,
		HTTPClient:      options.HTTPClient,
		ProtocolVersion: options.ProtocolVersion,
		SessionID:       options.SessionID,
		onMessage:       options.OnMessage,
	}
	instance.nextRequestID.Store(2)
	if instance.onMessage == nil {
		instance.onMessage = func(context.Context, map[string]any) error { return nil }
	}
	return instance
}

func (session *Session) SetMessageHandler(onMessage MessageHandler) {
	if onMessage == nil {
		session.onMessage = func(context.Context, map[string]any) error { return nil }
		return
	}
	session.onMessage = onMessage
}

func (session *Session) SendEnvelope(
	ctx context.Context,
	body map[string]any,
	onMessage MessageHandler,
) (RPCExchange, error) {
	handler := session.onMessage
	if onMessage != nil {
		handler = onMessage
	}
	return transport.PostRPC(
		ctx,
		session.HTTPClient,
		session.BaseURL,
		session.AuthToken,
		session.SessionID,
		session.ProtocolVersion,
		body,
		handler,
	)
}

func (session *Session) Request(
	ctx context.Context,
	method string,
	params map[string]any,
	onMessage MessageHandler,
) (RPCExchange, error) {
	body := map[string]any{
		"jsonrpc": "2.0",
		"id":      session.nextRequestID.Add(1) - 1,
		"method":  method,
	}
	if params != nil {
		body["params"] = params
	}
	return session.SendEnvelope(ctx, body, onMessage)
}

func (session *Session) RequestResult(
	ctx context.Context,
	method string,
	params map[string]any,
	onMessage MessageHandler,
) (map[string]any, error) {
	exchange, err := session.Request(ctx, method, params, onMessage)
	if err != nil {
		return nil, err
	}
	message, err := transport.TerminalMessage(exchange.Messages, exchange.Request["id"])
	if err != nil {
		return nil, err
	}
	result, ok := message["result"].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("terminal response did not include an object result")
	}
	return result, nil
}

func (session *Session) Notification(
	ctx context.Context,
	method string,
	params map[string]any,
	onMessage MessageHandler,
) (RPCExchange, error) {
	handler := session.onMessage
	if onMessage != nil {
		handler = onMessage
	}
	body := map[string]any{
		"jsonrpc": "2.0",
		"method":  method,
	}
	if params != nil {
		body["params"] = params
	}
	return transport.PostNotification(
		ctx,
		session.HTTPClient,
		session.BaseURL,
		session.AuthToken,
		session.SessionID,
		session.ProtocolVersion,
		body,
		handler,
	)
}

func (session *Session) ListTools(ctx context.Context) (map[string]any, error) {
	return session.RequestResult(ctx, "tools/list", map[string]any{}, nil)
}

func (session *Session) CallTool(ctx context.Context, name string, arguments map[string]any) (map[string]any, error) {
	return session.RequestResult(ctx, "tools/call", map[string]any{
		"name":      name,
		"arguments": arguments,
	}, nil)
}

func (session *Session) ListResources(ctx context.Context) (map[string]any, error) {
	return session.RequestResult(ctx, "resources/list", map[string]any{}, nil)
}

func (session *Session) ReadResource(ctx context.Context, uri string) (map[string]any, error) {
	return session.RequestResult(ctx, "resources/read", map[string]any{"uri": uri}, nil)
}

func (session *Session) SubscribeResource(ctx context.Context, uri string) (map[string]any, error) {
	return session.RequestResult(ctx, "resources/subscribe", map[string]any{"uri": uri}, nil)
}

func (session *Session) UnsubscribeResource(ctx context.Context, uri string) (map[string]any, error) {
	return session.RequestResult(ctx, "resources/unsubscribe", map[string]any{"uri": uri}, nil)
}

func (session *Session) ListResourceTemplates(ctx context.Context) (map[string]any, error) {
	return session.RequestResult(ctx, "resources/templates/list", map[string]any{}, nil)
}

func (session *Session) ListPrompts(ctx context.Context) (map[string]any, error) {
	return session.RequestResult(ctx, "prompts/list", map[string]any{}, nil)
}

func (session *Session) GetPrompt(ctx context.Context, name string, arguments map[string]any) (map[string]any, error) {
	params := map[string]any{"name": name}
	if arguments != nil {
		params["arguments"] = arguments
	}
	return session.RequestResult(ctx, "prompts/get", params, nil)
}

func (session *Session) Complete(ctx context.Context, params map[string]any) (map[string]any, error) {
	return session.RequestResult(ctx, "completion/complete", params, nil)
}

func (session *Session) SetLogLevel(ctx context.Context, level string) (map[string]any, error) {
	return session.RequestResult(ctx, "logging/setLevel", map[string]any{"level": level}, nil)
}

func (session *Session) ListTasks(ctx context.Context) (map[string]any, error) {
	return session.RequestResult(ctx, "tasks/list", map[string]any{}, nil)
}

func (session *Session) GetTask(ctx context.Context, taskID string) (map[string]any, error) {
	return session.RequestResult(ctx, "tasks/get", map[string]any{"taskId": taskID}, nil)
}

func (session *Session) GetTaskResult(ctx context.Context, taskID string) (map[string]any, error) {
	return session.RequestResult(ctx, "tasks/result", map[string]any{"taskId": taskID}, nil)
}

func (session *Session) CancelTask(ctx context.Context, taskID string) (map[string]any, error) {
	return session.RequestResult(ctx, "tasks/cancel", map[string]any{"taskId": taskID}, nil)
}

func (session *Session) Close(ctx context.Context) (DeleteSessionResult, error) {
	return transport.DeleteSession(
		ctx,
		session.HTTPClient,
		session.BaseURL,
		session.AuthToken,
		session.SessionID,
	)
}
